File: Frequency Distribution/Java/FrequencyDistributionMap.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.IntSummaryStatistics;
import java.util.List;

public class FrequencyDistributionMap {

    private int[] data;
    private int size;
    private int min;
    private int max;
    private int range;
    private Long classAmount;
    private int classLength;

    private List<Integer> _intervalBottom = new ArrayList<>();
    private List<Integer> _intervalTop = new ArrayList<>();
    private List<Integer> frequency = new ArrayList<>();
    private List<Double> classMark = new ArrayList<>();
    private List<Double> trueLowerLimit = new ArrayList<>();
    private List<Double> trueUpperLimit = new ArrayList<>();
    private List<Integer> cumulativeFrequencyLessThanType = new ArrayList<>();
    private List<Integer> cumulativeFrequencyGreaterThanType = new ArrayList<>();

    public FrequencyDistributionMap(int[] data){
        this.data = data;
        size = data.length;
        IntSummaryStatistics stat = Arrays.stream(data).summaryStatistics();
        min = stat.getMin();
        max = stat.getMax();
        range = max - min;
        classAmount = Math.round(1+((3.3)*Math.log10(data.length)));
        classLength = range / Math.round(classAmount);
        setupInterval();
        setupFrequency();
        setupLessAndGreater();
    }

    public String getInterval(int index, String splitter){
        return _intervalBottom.get(index) + splitter + _intervalTop.get(index);
    }

    public int getIntervalBottom(int index){
        return _intervalBottom.get(index);
    }

    public int getIntervalTop(int index){
        return _intervalTop.get(index);
    }

    public int getFrequency(int index){
        return frequency.get(index);
    }

    public double getClassMark(int index){
        return classMark.get(index);
    }

    public String getClassBoundaries(int index, String splitter){
        return trueLowerLimit.get(index) + splitter + trueUpperLimit.get(index);
    }

    public double getTrueLowerLimit(int index){
        return trueLowerLimit.get(index);
    }

    public double getTrueUpperLimit(int index){
        return trueUpperLimit.get(index);
    }

    public int getCumulativeFrequencyLessThanType(int index){
        return cumulativeFrequencyLessThanType.get(index);
    }

    public int getCumulativeFrequencyGreaterThanType(int index){
        return cumulativeFrequencyGreaterThanType.get(index);
    }

    public int getRow(){
        return _intervalBottom.size();
    }

    public int getSize(){
        return size;
    }

    public int getMin(){
        return min;
    }

    public int getMax(){
        return max;
    }

    public Double getMean(){
        Double total = 0.0;
        for (int i = 0; i < classAmount; i++){
            total = total + (frequency.get(i) * classMark.get(i));
        }
        return total / getSize();
    }

    public int getRange(){
        return range;
    }

    public Long getClassAmount(){ return classAmount; }

    public int getClassLength(){ return classLength; }

    private void setupInterval(){
        for (int i = 0; i < classAmount; i++){
            addInterval((int) (getMin() + ((getClassLength() + 1)*i)), (int) ((getMin() + getClassLength()) + ((getClassLength() + 1)*i)));
        }
    }

    private void setupFrequency(){
        for (int i = 0; i < getSize(); i++){
            for (int j = 0; j < classAmount; j++){
                if (data[i] >= _intervalBottom.get(j) && data[i] <= _intervalTop.get(j)){
                    int count = frequency.get(j);
                    count++;
                    frequency.set(j,count);
                }
            }
        }
    }

    private void setupLessAndGreater(){
        int countLess = 0;
        int countGreater = getSize();
        for (int i = 0; i < classAmount; i++){
            countLess += frequency.get(i);
            cumulativeFrequencyLessThanType.add(countLess);
            countGreater -= frequency.get(i);
            cumulativeFrequencyGreaterThanType.add(countGreater);
        }
    }

    private void addInterval(int intervalBottom, int intervalTop){
        _intervalBottom.add(intervalBottom);
        _intervalTop.add(intervalTop);
        frequency.add(0);
        classMark.add((double) intervalBottom + (((double) intervalTop - (double) intervalBottom) / 2));
        trueLowerLimit.add((double) intervalBottom - 0.5);
        trueUpperLimit.add((double) intervalTop + 0.5);
    }

}

File: Frequency Distribution/python/FrequencyDistributionMap.py
import math


class FrequencyDistributionMap:
    def __init__(self, data):
        self.data = data
        self.size = len(data)
        self.min = min(data)
        self.max = max(data)
        self.range = self.max - self.min
        self.classAmount = round(1 + (3.3 * math.log10(self.size)))
        self.classLength = round(self.range / round(self.classAmount)) - 1
        self._intervalBottom = []
        self._intervalTop = []
        self.frequency = []
        self.classMark = []
        self.trueLowerLimit = []
        self.trueUpperLimit = []
        self.cumulativeFrequencyLessThanType = []
        self.cumulativeFrequencyGreaterThanType = []
        self.setupInterval()
        self.setupFrequency()
        self.setupLessAndGreater()

    def getInterval(self, index, splitter):
        return str(self._intervalBottom[index]) + splitter + str(self._intervalTop[index])

    def getIntervalBottom(self, index):
        return self._intervalBottom[index]

    def getIntervalTop(self, index):
        return self._intervalTop[index]

    def getFrequency(self, index):
        return self.frequency[index]

    def getClassMark(self, index):
        return self.classMark[index]

    def getClassBoundaries(self, index, splitter):
        return str(self.trueLowerLimit[index]) + splitter + str(self.trueUpperLimit[index])

    def getTrueLowerLimit(self, index):
        return self.trueLowerLimit[index]

    def getTrueUpperLimit(self, index):
        return self.trueUpperLimit[index]

    def getCumulativeFrequencyLessThanType(self, index):
        return self.cumulativeFrequencyLessThanType[index]

    def getCumulativeFrequencyGreaterThanType(self, index):
        return self.cumulativeFrequencyGreaterThanType[index]

    def getRow(self):
        return len(self._intervalBottom)

    def getSize(self):
        return self.size

    def getMin(self):
        return self.min

    def getMax(self):
        return self.max

    def getMean(self):
        total = 0.0
        for i in range(0, self.classAmount):
            total += (self.frequency[i] * self.classMark[i])
        return total / self.getSize()

    def getRange(self):
        return self.range

    def getClassAmount(self):
        return self.classAmount

    def getClassLength(self):
        return self.classLength

    def setupInterval(self):
        for i in range(0, self.classAmount):
            self.addInterval(self.getMin() + ((self.getClassLength() + 1) * i),
                             (self.getMin() + self.getClassLength()) + ((self.getClassLength() + 1) * i))

    def setupFrequency(self):
        for i in range(0, self.getSize()):
            for j in range(0, self.classAmount):
                if self._intervalBottom[j] <= self.data[i] <= self._intervalTop[j]:
                    count = self.frequency[j] + 1
                    self.frequency[j] = count

    def setupLessAndGreater(self):
        countLess = 0
        countGreater = self.getSize()
        for i in range(0, self.classAmount):
            countLess += self.frequency[i]
            self.cumulativeFrequencyLessThanType.append(countLess)
            countGreater -= self.frequency[i]
            self.cumulativeFrequencyGreaterThanType.append(countGreater)

    def addInterval(self, intervalBottom, intervalTop):
        self._intervalBottom.append(intervalBottom)
        self._intervalTop.append(intervalTop)
        self.frequency.append(0)
        self.classMark.append(intervalBottom + ((intervalTop - intervalBottom) / 2))
        self.trueLowerLimit.append(intervalBottom - 0.5)
        self.trueUpperLimit.append(intervalTop + 0.5)
